cudax/include/cuda/experimental/__stf/internal/parallel_for_scope.cuh

```diff
@@ -74,24 +77,330 @@ __global__ void loop(const _CCCL_GRID_CONSTANT size_t n, shape_t shape, F f, tup
+    // TODO add is invocable test on type
+    ElementType::apply_op(::std::get<idx>(tup), ::std::get<idx>(src));
+  }
+
+  // Main function: applies apply_op to every element of the tuple
+  template <typename Tuple, ::std::size_t... Is>
+  __device__ void apply_op_impl(const Tuple& src, ::std::index_sequence<Is...>)
+  {
+    // Expand the indices and call apply_op_impl_i for each index
+    (apply_op_impl_i<Is>(src), ...);
+  }
+
+  // Helper function to select and return the correct element
+  template <size_t Is>
+  __device__ decltype(auto) select_element(tuple_args& targs)
+  {
+    using OpType = typename ::std::tuple_element<Is, tuple_ops>::type;
+    if constexpr (::std::is_same_v<OpType, task_dep_op_none>)
+    {
+      return ::std::get<Is>(targs); // Return reference to targs[i]
+    }
+    else
+    {
+      return ::std::get<Is>(tup); // Return reference to redux_buffer[i]
+    }
+  }
+
+  template <::std::size_t... Is>
+  __device__ auto make_targs_aux_impl(tuple_args& targs, ::std::index_sequence<Is...>)
+  {
+    // We do not use make_tuple to preserve references
+    return ::std::forward_as_tuple(select_element<Is>(targs)...);
+  }
+
+  // Call the init_op method if this is reduction operator
+  template <size_t Is>
+  __device__ decltype(auto) init_element()
+  {
+    using OpType = typename ::std::tuple_element<Is, tuple_ops>::type;
+    if constexpr (!::std::is_same_v<OpType, task_dep_op_none>)
+    {
+      OpType::init_op(::std::get<Is>(tup));
+    }
+  }
+
+  template <::std::size_t... Is>
+  __device__ void init_impl(::std::index_sequence<Is...>)
+  {
+    (init_element<Is>(), ...);
+  }
+
+  /* This tuple contains either EmptyType for non reduction variables, or the owning type for a reduction variable.
+   * if tuple_args = tuple<slice<double>, scalar<int>, slice<int>> and tuple_ops=tuple<none, sum<int>, none>
+   * this will correspond to tuple<EmptyType, int, EmptyType>.
+   *
+   * So we can store that tuple in shared memory to perform per-block reduction operations.
+   */
+  redux_buffer_tup_t<tuple_args, tuple_ops> tup;
+};
+
+/* the redux_buffer is an array of tuples which sizes corresponds to the number of CUDA blocks */
+template <typename F, typename shape_t, typename tuple_args, typename tuple_ops>
+__global__ void loop_redux(
+  const _CCCL_GRID_CONSTANT size_t n,
+  shape_t shape,
+  F f,
+  tuple_args targs,
+  redux_buffer_tup_t<tuple_args, tuple_ops>* redux_buffer)
+{
+  size_t i          = blockIdx.x * blockDim.x + threadIdx.x;
+  const size_t step = blockDim.x * gridDim.x;
+
+  // This tuple in shared memory contains either an empty type for "regular"
+  // arguments, or an owning local variable for reduction variables.
+  // extern __shared__ redux_buffer_tup_wrapper<tuple_args, tuple_ops> per_block_redux_buffer[];
+  extern __shared__ char dyn_buffer[];
+  auto* per_block_redux_buffer = (redux_buffer_tup_wrapper<tuple_args, tuple_ops>*) ((void*) dyn_buffer);
```

Comment:
this weirdness is due to the fact that external symbols won't work with different type for the same symbol

Reply:
Some comment was added to explain it